### PR TITLE
Add functionality to optionally enable animation method calls in the editor.

### LINKED ISSUE #9939: Allow Animation Method Call tracks to be optionally enabled in the editor.
### Describe the project you are working on

A project that uses AnimationPlayer to call methods on particle systems and audio players. 

### Describe the problem or limitation you are having in your project

Animation Method Calls do not get run in the editor for safety. 

This makes it very painful to work with particle systems (calling GPUParticles3D.Restart() at the start of the animation doesn't work in the editor).
Also, AudioStreamPlayers (calling AudioStreamPlayers.Play() to be able to use AudioStreamPlayers.max_polyphony), doesn't work in the editor.

This all slows down my workflow as I have to play the game each time I want to see my changes. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add an option to enable method calls in the editor, per key. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

img

### If this enhancement will not be used often, can it be worked around with a few lines of script?

 Technically yes, but tedious. I have to use a hack using a @tool script with a setter on a property to call a method in the editor.

### Is there a reason why this should be core and not an add-on in the asset library?

Animation is core.

### PR DESCRIPTION
**Pull request by Ryan-000**
Resolves https://github.com/godotengine/godot-proposals/issues/9939

Related: https://github.com/godotengine/godot/issues/21388, https://github.com/godotengine/godot/issues/75479, https://github.com/godotengine/godot/issues/73043

### CODE REVIEW

**Comment by Helpsypoo** [comment]
Is this still alive? Wanted to add support for it.

For what it's worth, I've been hacking around this by animating properties with setters that run my desired code. It's ugly, but it's the only way to preview my method call animations with the editor.

**Review by TokageItLab — CHANGES_REQUESTED** [review]
We have explained many times in the past that implicitly calling methods on the editor is not allowed as it can easily break the editor and should never be approached this way as it is also a security vulnerability.

If it is to be supported in the core, it must implement some kind of explicit API for registration, similar to the way some plugins are registered.

**Comment by Helpsypoo** [comment]
Thanks for your reply. Is there a go-to place documenting the details here? I looked through related issues before commenting, but I don't quite understand. 

I recognize you don't owe me a full explanation, but I'll ask in case you think answering will help others better understand this (since it seems perennial):

What is the difference between calling a method from an animation and calling a method from a tool script? Tool scripts can also easily break the editor. On the security front, is it that someone might share a project that calls, for example, a SendMeYourBankAccountInfo method, and it would be hard to notice that if it's buried deeply in an animation?

**Comment by TokageItLab** [comment]
> After all, method tracks already execute methods when the game is run. If running these methods in the editor is considered risky, how is it fundamentally different from them being executed during runtime? 

As I said above, the risk is that it could be executed the moment the project is opened. You can look at the code before running the game to see if there is any malicious code in the game.

Indeed, there is a current convenience concern, but in the rush to resolve it, we will never allow a dangerous Hack instead of appropriate PR. Removing the Hack in the future will break compatibility and will not be a good experience for both maintainers and users.

**Comment by Helpsypoo** [comment]
> I don't know why you want to use Method Tracks on your editor

My use case is admittedly somewhat niche. I make simulation videos ([here's one made with godot](https://youtu.be/tCoEYFbDVoI?si=P6Y8Xe4hLSl-b3Hq)). I write code to construct animations which are combined together into one large animation with animation playback tracks. Since the scenes can be intricate, it's often convenient to create method call tracks to schedule larger operations that kick off tweens. It's also useful in cases where I don't know everything ahead of time. For example, I currently have a simulation that runs in real time (which I'd like to put into a game at some point), and I have graphs showing the simulation data that I want to update at specific times to align with voiceover. Method calls allow these graphs to query the current simulation data at the time the animation is called. It's possible to time these without method call tracks, but method call tracks match the rest of my process, and being able to preview the graph animations for quick iteration in the editor is very useful (hence me resorting to the hack of animating properties with complex setters).

I've also worked on a jam game (in Unity) where we set up animation events to trigger code from inside the animation, which made it easy to dependably sync other object behaviors to various parts of animations (projectiles in our case).

Tldr, it's a very powerful feature for the animation system.

Thanks for this discussion. I understand better now why it's a concern. Thanks for giving me the opportunity to explain my use case, and for the care you put into developing the engine. ✌️

**Comment by GamplaxPramgrammer** [comment]
> Thank you for the feedback, @TokageItLab. I value the discussion around security concerns. However, I find it difficult to see how the proposed call_in_editor functionality introduces risks that aren’t already present elsewhere in the engine.
> 
> For example:
> 
>     * `@tool` GDScript in autoloads are executed automatically in the editor
> 
>     * Similarly, `@tool` scripts embedded or referenced in resources (e.g., .tres, .tscn, or .res files) are executed when those resources are loaded in the editor.
> 
>     * GDExtension libraries are automatically loaded and executed by the engine.
> 
> 
> Arguably, the proposed call_in_editor functionality introduces a clear, visible flag that explicitly marks which animation method calls are allowed to execute in the editor. This seems more transparent and controlled than many of the existing mechanisms, which operate implicitly without user awareness.
> 
> Additionally, if security is a primary concern, it raises a broader question about the method call track feature as a whole. After all, method tracks already execute methods when the game is run. If running these methods in the editor is considered risky, how is it fundamentally different from them being executed during runtime? If this is truly a concern, then perhaps the entire method call track system needs to be reconsidered rather than singling out the editor use case.
> 
> As it is right now, if your animation depends on having methods being run (and this could be more than just for GPU particles), you are forced to restart the game each time to see the effect of your changes. On top of that, you also need to set up a test scene with a camera in it just to preview the animation, which makes the animation workflow pretty difficult and cumbersome when faster iteration is desired.
> 
> I hope this perspective helps clarify the intent behind the proposal and provides some useful context for further discussion. I’m more than happy to explore alternatives if they are still needed to address any outstanding concerns. Thanks again for your insights!

One thing I think that proposed functionality would help in my scenario.
Currently, VideoStreamPlayer doesn't have a "playing" attribute like AudioStreamPlayer does, making it extremely difficult to preview the videos in the editor in relation to using Animation Player to do additional functionality alongside playing the video. As a work-around, I used a bunch of `@tool` scripts.
But having that functionality would let me call the function to play the video.
